### DISCUSSION TITLE
on_failure_callback getting the wrong context

### DISCUSSION THREAD

**Discussion by freyrsae** [apache/airflow#24991] [General]
### Apache Airflow version

2.2.5

### What happened

We have slack alerting on failure enabled in our DAG and use `context.get('task_instance')` to customise the error message but it is usually the wrong task_instance we get. Seems very familiar to https://github.com/apache/airflow/discussions/19371

### What you think should happen instead

I should get the task instance of the task that failed, not just some task in the dag.

### Operating System

Running with Composer

### Deployment

Composer

### Deployment details

Using composer-1.19.2-airflow-2.2.5

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
If you want to open an issue, you must provide some evidences, that we will be able to look at and confirm and diagnose the issue. Example DAG, logs, screenshots from the UI, emails that you received and corresponding task logs. This is might be something connected with your (mis)configuration, or the way your DAGs are written. There is no way anyone could do anything with such a "bare" report you've done without any of those. Please spend your time on gathering the evidences and presenting it in the way that it will be easy to understand and diagnose the error (and see if it's not you who made some mistakes).

This is the best way you can contribute back for the free software you got - remember that most people here helps in their free time, so you show a bit disrespect to their time if you do not spend time on reporting more details that will help them to analyse it. 

Please [Exercise your empathy](https://www.youtube.com/watch?v=G6VjYvKr2wQ&list=PLGudixcDaxY2LxjeHpZRtzq7miykjjFOn&index=55)

Converted it into discussion/

**Comment by hussein-awala** [comment]
I think there is something unclear with this argument:
```python
import pendulum

from airflow.operators.python import PythonOperator
from airflow.models import DAG


def failure_callback(context):
    with open(f"/opt/airflow/failed_task.txt", "a") as f:
        f.write(f"{context['ti'].task_id}: {context['ti'].state}\n")


with DAG(
    dag_id='test_dag',
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    schedule=None,
    on_failure_callback=failure_callback
) as dag:

    task1 = PythonOperator(
        task_id="task1",
        python_callable=lambda: print("ok")
    )

    task2 = PythonOperator(
        task_id="task2",
        python_callable=lambda: 1/0
    )

    task3 = PythonOperator(
        task_id="task3",
        python_callable=lambda: print("ok")
    )

    task4 = PythonOperator(
        task_id="task4",
        python_callable=lambda: print("ok")
    )

    task5 = PythonOperator(
        task_id="task5",
        python_callable=lambda: print("ok")
    )

    task1 >> task2 >> task3 >> task4
    task2 >> task5

```
![image](https://user-images.githubusercontent.com/21311487/216198682-8d889d1d-c082-4805-be22-29c7607bb8f7.png)

And when I check the file:
```bash
root@3ee0ea70f557:/opt/airflow# cat failed_task.txt
task5: upstream_failed 
```
So the callback is called with context of the last defined task, regardless its state:
```python
import pendulum

from airflow.operators.python import PythonOperator
from airflow.models import DAG
from airflow.utils.trigger_rule import TriggerRule


def failure_callback(context):
    with open(f"/opt/airflow/failed_task.txt", "a") as f:
        f.write(f"{context['ti'].task_id}: {context['ti'].state}\n")


with DAG(
    dag_id='test_dag',
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    schedule=None,
    on_failure_callback=failure_callback
) as dag:

    task1 = PythonOperator(
        task_id="task1",
        python_callable=lambda: print("ok")
    )

    task2 = PythonOperator(
        task_id="task2",
        python_callable=lambda: 1/0
    )

    task3 = PythonOperator(
        task_id="task3",
        python_callable=lambda: print("ok")
    )

    task4 = PythonOperator(
        task_id="task4",
        python_callable=lambda: print("ok")
    )

    task5 = PythonOperator(
        task_id="task5",
        python_callable=lambda: print("ok"),
        trigger_rule=TriggerRule.ALL_DONE,
    )

    task1 >> task2 >> task3 >> task4
    task2 >> task5
```
![image](https://user-images.githubusercontent.com/21311487/216198764-c0e4aa16-3aef-4a5d-84cf-0f11261d6ccc.png)

This create the file:
```bash
root@3ee0ea70f557:/opt/airflow# cat failed_task.txt
task5: success
```

I think the goal of the `on_failure_callback` on the Dag level, is providing the context of the `DagRun` which is identical between the tis, for that we don't care which ti context we provide. But this is not clear in the doc, so we need to clarify this point.

An alternative solution to run the callback on each failed ti, is providing it as a default arg:
```python
with DAG(
    dag_id='test_dag',
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    schedule=None,
    default_args=dict(on_failure_callback=failure_callback)
) as dag:
```

**Reply by josh-fell** [reply]
>I think the goal of the on_failure_callback on the Dag level, is providing the context of the DagRun which is identical between the tis, for that we don't care which ti context we provide. But this is not clear in the doc, so we need to clarify this point.

Yes. I've assigned myself #26760 to address this and other things.

**Reply by Taragolis** [reply]
Just FYI, there is PR for "fix" this issue exists
I've write down the reason why it happen (https://github.com/apache/airflow/pull/27616#discussion_r1093536618), and what other options (potentially more horrible) we have right now

**Reply by Taragolis** [reply]
Oh... ignore me, I've talk about Task level callback, but this discussion about DAG Level 🤦

**Reply by hussein-awala** [reply]
@josh-fell thank you! And sorry, I didn't see the issue you mentioned.

**Comment by josh-fell** [comment]
Previously mentioned in this issue as well: https://github.com/apache/airflow/issues/26760
